### PR TITLE
[Scala] More Optimizers

### PR DESCRIPTION
**Pull request by terrytangyuan**
@javelinjs @tqchen Could any of you help double check this?

### CODE REVIEW

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/AdaDelta.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+import ml.dmlc.mxnet.NDArrayConversions._
+
+/**
+  * AdaDelta optimizer as described in Matthew D. Zeiler, 2012.
+  * http://arxiv.org/abs/1212.5701
+  *
+  * <b>WARNING</b>
+  * TODO: This class has NOT been tested yet.
+  * And there exists potential <b>memory leak</b> in the implementation
+  *
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/AdaDelta.scala]: Please fix these two issues, so as the Adam optimizer.

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/AdaGrad.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  * And there exists potential <b>memory leak</b> in the implementation
+  *
+  * @author Yuan Tang
+  *
+  * @param learningRate Float, Step size.
+  * @param epsilon Float
+  * @param rescaleGradient Float, rescaling factor of gradient.
+  * @param wd Float, L2 regularization coefficient add to all the weights
+  */
+class AdaGrad(var learningRate: Float = 0.05f, val rescaleGradient: Float = 1.0f,
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/AdaGrad.scala]: can learningRate be val?

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/AdaDelta.scala (hunk 1 above)
Comment (terrytangyuan) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/AdaDelta.scala]: Thanks. I will fix them later today.

**Comment by javelinjs** [comment]
@terrytangyuan Please don't merge without others' confirmation. There's still problems in this PR, I'll help to fix. Be cautious next time, thanks.

File: scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/AdaDelta.scala [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      resdGrad = NDArray.clip(resdGrad, -clipGradient, clipGradient)
+      oldResdGrad.dispose()
+    }
+
+    val (acc_g, acc_delta) = state.asInstanceOf[(NDArray, NDArray)]
+
+    acc_g.set(this.rho * acc_g + (1.0f - this.rho) * resdGrad * resdGrad)
+    val current_delta = NDArray.sqrt(acc_delta + this.epsilon) /
+      NDArray.sqrt(acc_g + this.epsilon) * resdGrad
+    acc_delta.set(this.rho * acc_delta + (1.0f - this.rho) * current_delta * current_delta)
```
Comment (javelinjs) [inline scala-package/core/src/main/scala/ml/dmlc/mxnet/optimizer/AdaDelta.scala]: middle result not disposed; variable name convention

**Comment by terrytangyuan** [comment]
Thanks. I'll be quite busy on the following couple of days so was thinking to merge it first. Thanks for double checking and help!

**Pull request merged**